### PR TITLE
Specify names on service update and generate names client-side.

### LINKED ISSUE #277: Allow automatic client-side revision name generation
## In what area(s)?
Workflow and client conventions

## Describe the feature:

Previously, in the client-conventions.md (still in serving, still meaning to port it over here), we suggested adding a nonce to force creation of a new revision and to make that new revision findable once you'd created it.

Now that v1beta1 style (even in v1alpha1) allows for bring-your-own-revision-name, it's much more natural to autogenerate a name for the revision on the client side. We should support random name generation on the client side, and have it on by default, so that the revisions are findable, at least, and so that when you deploy an image by a tag that's changed it forces creation of a new revision.

### PR DESCRIPTION
**Pull request by sixolet**
Fixes #277 #276 

## Proposed Changes

* `--revision-name` flag allows setting revision names in service update. If you don't provide the service name as a prefix, it'll get auto-added for you.
* `--generate-revision-name` flag (default TRUE) makes revision names client side by default.

Whether to include the generation is up for discussion. Otherwise this is substantially similar to the server-side generateName(), up to and including our lack of vowels.

**Release Note**

```release-note

```

### CODE REVIEW

**Comment by sixolet** [comment]
Cool, now that I've dealt with the old-api cases I feel happier about this.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 1]
```diff
@@ -85,6 +127,23 @@ func (p *ConfigurationEditFlags) Apply(service *servingv1alpha1.Service, cmd *co
 		return err
 	}
 
+	name := ""
```
Comment (rhuss) [inline pkg/kn/commands/service/configuration_edit_flags.go]: This `apply()` method is already quite a mouthful could this revision name handling block be extracted in an own method e.g. `setRevisionName(cmd)` ?

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 2]
```diff
@@ -26,51 +26,93 @@ import (
...
+	flags []string
 }
 
 type ResourceFlags struct {
 	CPU    string
 	Memory string
 }
 
-func (p *ConfigurationEditFlags) AddUpdateFlags(command *cobra.Command) {
+func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
```
[the remark this answers is not in the document]
Comment (duglin) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Can you add a comment to explain that "Shared" means that these flags are common between both create() and update() ?

[the remark this answers is not in the document]
Comment (duglin) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Let's get this PR in before #345 so that #345 can add its flags in here too!

File: pkg/serving/service.go [hunk 3]
```diff
@@ -38,6 +41,23 @@ func RevisionTemplateOfService(service *servingv1alpha1.Service) (*servingv1alph
 	return config.DeprecatedRevisionTemplate, nil
 }
 
+var charChoices = []string{
+	"b", "c", "d", "f", "g", "h", "j", "k", "l", "m", "n", "p", "q", "r", "s", "t", "v", "w", "x",
```
Comment (duglin) [inline pkg/serving/service.go]: I gotta ask... did you purposely remove some chars to reduce the chances of inappropriate words?  :-)

Comment (rhuss) [reply pkg/serving/service.go]: @duglin see initial PR comment:

> Otherwise this is substantially similar to the server-side generateName(), up to and including our lack of vowels.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 4]
```diff
@@ -85,6 +127,23 @@ func (p *ConfigurationEditFlags) Apply(service *servingv1alpha1.Service, cmd *co
...
+		name = p.RevisionName
+		if !strings.HasPrefix(name, service.Name+"-") {
+			name = service.Name + "-" + name
+		}
+	} else if p.GenerateRevisionName {
+		name = servinglib.GenerateRevisionName(service)
+	}
+	if anyMutation {
+		err = servinglib.UpdateName(template, name)
+		if err == servinglib.ApiTooOldError && !cmd.Flags().Changed("revision-name") {
```
[the remark this answers is not in the document]
Comment (rhuss) [reply pkg/kn/commands/service/configuration_edit_flags.go]: I'm fine with both. It's not hard to understand, just unconventional for my eyes (not a Python guy, and in Perlland we never used such an idiom, we had better ways to obfuscate ;-P

File: pkg/kn/commands/service/service_update_test.go [hunk 5]
```diff
@@ -398,6 +491,23 @@ func newEmptyService() *v1alpha1.Service {
 	}
 }
 
+func newEmptyServiceBeta() *v1alpha1.Service {
```
Comment (rhuss) [inline pkg/kn/commands/service/service_update_test.go]: doesnt look very 'beta'

File: pkg/serving/service.go [hunk 6]
```diff
@@ -38,6 +42,40 @@ func RevisionTemplateOfService(service *servingv1alpha1.Service) (*servingv1alph
...
+	chars := make([]string, 0, l)
+	for i := 0; i < l; i++ {
+		chars = append(chars, charChoices[rand.Int()%len(charChoices)])
+	}
+	return strings.Join(chars, "")
+}
+
+// GenerateRevisionName returns an automatically-generated name suitable for the
+// next revision of the given service.
+func GenerateRevisionName(nameTempl string, service *servingv1alpha1.Service) (string, error) {
```
Comment (rhuss) [inline pkg/serving/service.go]: Can we add some tests for this method ? (and maybe also when given as argument from outside in service_update_test ?)

**Comment by rhuss** [comment]
If I understand the motivation of client-side generation as described in https://github.com/knative/client/blob/master/conventions/client.md#change-non-code-attributes then it's about the potential risk of a race condition so that `latestCreatedRevisionName` could not be used for obtaining the image digest.

I doubt a bit that this could be the case as you always have an optimistic lock on the service resource, so when two clients simultaneously try to update the service the first one wins, and the second one has to re-fetch the resource (with the updated `latestCreatedRevisionName`). I'm not sure whether this update of the latestCreatedRevisionName is done in an admission controller or async. 

In the latter case, there is still a slight chance for a race, but is this really that severe?

* Two clients have to operate at the same time
* The digest behind the image tag must have been changed. This is likely the case for `latest`. It could be the case for a version like tag (`1.2`) but although possible and sometimes done, I consider this a very unlikely event.

So if there should be a race for `latest` which this might result in an hick-up of a development workflow (as `latest` is not used in production).

I don't question the algorithm in the document at all, it has been the result of a lengthy (and probably painful) discussion. I just want to point out that I personally would take that risk and just reuse `lastesCreatedRevisionName` always for finding the active revision, regardless of who has created the name (since we are in an update or change situation we will create a new revision anyway, no need for a trigger in this case)

**Comment by sixolet** [comment]
`latestCreatedRevisionName` is updated async, which is the race opportunity, yeah.

The chance of the race is low, but the consequences can be high (running code you specifically did not intend to run)

File: pkg/kn/commands/service/service_update_test.go (hunk 5 above)
Comment (sixolet) [reply pkg/kn/commands/service/service_update_test.go]: Renamed.

File: pkg/serving/service.go (hunk 6 above)
Comment (sixolet) [reply pkg/serving/service.go]: Done.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 7]
```diff
@@ -85,6 +127,23 @@ func (p *ConfigurationEditFlags) Apply(service *servingv1alpha1.Service, cmd *co
...
 	}
 
+	name := ""
+	if cmd.Flags().Changed("revision-name") {
+		name = p.RevisionName
+		if !strings.HasPrefix(name, service.Name+"-") {
+			name = service.Name + "-" + name
+		}
+	} else if p.GenerateRevisionName {
+		name = servinglib.GenerateRevisionName(service)
```
[the remark this answers is not in the document]
Comment (sixolet) [reply pkg/kn/commands/service/configuration_edit_flags.go]: The server will reject. I think this'll work as designed with people able to template their own.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 8]
```diff
@@ -85,6 +127,23 @@ func (p *ConfigurationEditFlags) Apply(service *servingv1alpha1.Service, cmd *co
...
+	}
+	if anyMutation {
```
[the remark this answers is not in the document]
Comment (sixolet) [reply pkg/kn/commands/service/configuration_edit_flags.go]: done

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 1 above)
Comment (sixolet) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Done

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 9]
```diff
@@ -26,45 +26,90 @@ import (
...
+	command.Flags().IntVar(&p.ConcurrencyTarget, "concurrency-target", 0,
+		"Recommendation for when to scale up based on the concurrent number of incoming request. "+
+			"Defaults to --concurrency-limit when given.")
+	p.markFlagMakesRevision("concurrency-target")
+	command.Flags().IntVar(&p.ConcurrencyLimit, "concurrency-limit", 0,
+		"Hard Limit of concurrent requests to be processed by a single replica.")
+	p.markFlagMakesRevision("concurrency-limit")
 	command.Flags().Int32VarP(&p.Port, "port", "p", 0, "The port where application listens on.")
+	p.markFlagMakesRevision("port")
+	command.Flags().StringVar(&p.RevisionName, "revision-name", "{{.Service}}-{{.Random 5}}-{{.Generation}}",
```
Comment (duglin) [inline pkg/kn/commands/service/configuration_edit_flags.go]: this seems excessively long :-)

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 10]
```diff
@@ -165,3 +224,14 @@ func (p *ConfigurationEditFlags) computeResources(resourceFlags ResourceFlags) (
 
 	return resourceList, nil
 }
+
+// AnyMutation returns tue if there are any revision template mutations in the
```
Comment (duglin) [inline pkg/kn/commands/service/configuration_edit_flags.go]: s/tue/true/

File: docs/cmd/kn_service_update.md [hunk 11]
```diff
@@ -41,6 +41,7 @@ kn service update NAME [flags]
   -p, --port int32               The port where application listens on.
       --requests-cpu string      The requested CPU (e.g., 250m).
       --requests-memory string   The requested memory (e.g., 64Mi).
+      --revision-name string     The revision name to set. Must start with the service name and a dash as a prefix. Empty revision name will result in the server generating a name for the revision. Accepts golang templates, allowing {{.Service}} for the service name, {{.Generation}} for the generation, and {{.Random [n]}} for n random consonants. (default "{{.Service}}-{{.Random 5}}-{{.Generation}}")
```
Comment (duglin) [inline docs/cmd/kn_service_update.md]: Since I suspect this flag will be used a lot, what do you think about having it available via `-r` too ?

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 9 above)
Comment (sixolet) [reply pkg/kn/commands/service/configuration_edit_flags.go]: It inspired #351 . I think it's better to describe clearly.

Comment (duglin) [reply pkg/kn/commands/service/configuration_edit_flags.go]: only the first time, after that when people do --help it'll just annoy them

File: docs/cmd/kn_service_update.md (hunk 11 above)
Comment (sixolet) [reply docs/cmd/kn_service_update.md]: Easy enough to do later.  Let's do the simpler thing for now — I'm nervous-er about doing shortopts w/o consensus much more than longopts.

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 9 above)
Comment (sixolet) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Hum. Roland suggested I _add_ the part about the empty string. I think the part about the template is valuable, as it is not discoverable otherwise. 

We could make it so, but we don't have a good place for that yet. I'd like to defer shortening it until we have a good place for other doc on options.

File: docs/cmd/kn_service_update.md (hunk 11 above)
Comment (rhuss) [reply docs/cmd/kn_service_update.md]: Now that I know the service name prefix is mandatory for a revision, I wonder whether we can help the user to do this for her automatically if not already existant? E.g. allowing for name `{{.Random 5}}-{{.Generation}}` (which would be a good benefit because of the verboseness of Go templates).

If not, we should at least add client-side validation for this constraint.

**Comment by sixolet** [comment]
If you pass the empty string, as the revision name, it's equivalent

**Comment by rhuss** [comment]
> If you pass the empty string, as the revision name, it's equivalent

Cool. So the last thing which needs clarification whether we enforce a `{{.Service}}-` prefix or allow everything and wait for the server-side error. I liked that this was added implicitly when not given.

**Comment by sixolet** [comment]
Oh I think I misread you as saying the opposite earlier. Easy enough!

**Comment by rhuss** [comment]
Yeah, sorry, totally missed that the service name as the prefix is a mandatory server-side requirement. Now that this is clear to me, I think we already should check and enforce that client-side. Apologies for my ignorance.

**Comment by sixolet** [comment]
Did you mean check and error, or fix it up for the user? I'm of two minds myself that's why I keep changing it back and forth

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 12]
```diff
@@ -26,45 +26,90 @@ import (
...
+		"The revision name to set. Must start with the service name and a dash as a prefix. "+
+			"Empty revision name will result in the server generating a name for the revision. "+
+			"Accepts golang templates, allowing {{.Service}} for the service name, "+
+			"{{.Generation}} for the generation, and {{.Random [n]}} for n random consonants.")
+	p.markFlagMakesRevision("revision-name")
 }
 
+// AddUpdateFlags adds the flags specific to update.
+func (p *ConfigurationEditFlags) AddUpdateFlags(command *cobra.Command) {
+	p.addSharedFlags(command)
```
Comment (navidshaikh) [inline pkg/kn/commands/service/configuration_edit_flags.go]: The better way to handle this separation would be to implement the suggestion from @cppforlife https://github.com/knative/client/pull/79#discussion_r283004004 logged as issue at https://github.com/knative/client/issues/110, that should go as separate PR though and shouldn't block this one.

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (navidshaikh) [reply pkg/kn/commands/service/configuration_edit_flags.go]: @duglin We add traffic flags separately on `service update` command. IMO they don't need to show up in `ConfigurationEditFlags` as updating traffic block is not configuration-edit operation and only traffic flags change don't result into a new revision.

**Comment by rhuss** [comment]
@sixolet I clearly would opt for fixing it for the user, as this is what the user would have to do anyway manually. Tbh I found it very annoying if a command knows how to fix sth, but doesn't do it automatically when there is a 99% certainty about the intention of the user.

My favourite (counter) example is:
```
$ find -name bla .
find: paths must precede expression: .
Usage: find [-H] [-L] [-P] [-Olevel] [-D help|tree|search|stat|rates|opt|exec] [path...] [expression]
```

(of course there could be ambiguities in certain cases, and then one can still throw an error. But for this very common 'user error' there should be an automatic "fix").

We could 'educate' the user by sending a warning if not `service-` prefixed, but its maybe also just fine to add this to the success message like

```
kn service create --image gcr.io/bla --revision-name="{{Rand 5}}-{{.Generation}}"
Service "myservice" with revision "myservice-dgnde-1" has been created.
```

and maybe something similar for `kn service update`.

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (duglin) [reply pkg/kn/commands/service/configuration_edit_flags.go]: From a UX perspective, I'd like to be able to setup the traffic during "service create" - that's all I'm asking for.

**Comment by duglin** [comment]
Seeing something odd:
```
$ ./kn service create echo --image duglin/echo --revision-name echo-v1
the service is using too old of an API format for the operation
```
But this works:
```
$ ./kn service create echo --image duglin/echo
Service 'echo' successfully created in namespace 'default'.
Waiting for service 'echo' to become ready ... OK

Service URL:
http://echo-default.kndev.us-south.containers.appdomain.cloud

$ ./kn service update echo --revision-name echo-v1
Waiting for service 'echo' to become ready ... OK
Service 'echo' updated in namespace 'default'.
```
I'm pretty sure I have both alpha1 and beta1 installed in my cluster.

**Comment by sixolet** [comment]
@duglin that's due to a related thing that I think should be fixed in a different PR: we default to creating services in the old api format, so a newly created service "is using the old api format" until it's created and then the server converts it.

**Comment by duglin** [comment]
ok, so when #361 is merged then all should be good, right?
btw - rebase needed on this one.

File: docs/cmd/kn_service_update.md (hunk 11 above)
Comment (duglin) [inline docs/cmd/kn_service_update.md]: Do we describe what "Generation" is in our docs some place?

**Comment by duglin** [comment]
/lgtm

Couple of comments/question though:
- I'm assuming once #361 goes in you'll create a PR to handle the `kn service create` variant, right?
- any reason we're no updating the docs in https://github.com/knative/client/blob/master/docs/cmd/kn_service_update.md with info about this?

Rebase needed.

**Comment by sixolet** [comment]
There we go. Rebased.
